drivers/epson/projector/esc_vp21.cr

```diff
@@ -0,0 +1,274 @@
+require "digest/md5"
+require "placeos-driver/interface/muteable"
+require "placeos-driver/interface/powerable"
+require "placeos-driver/interface/switchable"
+
+module Epson; end
+
+module Epson::Projector; end
+
+# Documentation: documentation link
+
+class Epson::Projector::EscVp21 < PlaceOS::Driver
+  include Interface::Powerable
+  include Interface::Muteable
+
+  enum Inputs
+    HDMI
+    HDBaseT
+  end
+
+  # Discovery Information
+  tcp_port 1024
+  descriptive_name "Epson Projectors"
+  generic_name :Display
+
+  def on_load
+    transport.tokenizer = Tokenizer.new("\r\n")
+
+    self[:volume_min] = 0
+    self[:volume_max] = 255
+
+    self[:power] = 0
+    self[:stable_state] = true
+
+    self[:type] = :projector
+  end
+
+  def on_update
+  end
+
+  def connected
+    # Have to init comms
+    send("ESC/VP.net\x10\x03\x00\x00\x00\x00")
+    do_poll
+    schedule.every(52.seconds) { do_poll }
+  end
+
+  def disconnected
+    self[:power] = false
+    schedule.clear
+
+    @channel.close unless @channel.closed?
+  end
+
+  # used to coordinate the projector password hash
+  @channel : Channel(String) = Channel(String).new
+
+  #
+  # Power commands
+  #
+  def power(state : Bool)
+    self[:stable_state] = false
+    if state
+      self[:power_target] = true
+      do_send(:PWR, true, timeout: 40000, name: :power)
+      logger.debug { "-- epson Proj, requested to power on" }
+      do_send(:PWR, name: :power_state)
+    else
+      self[:power_target] = false
+      do_send(:PWR, false, timeout: 10000, name: :power)
+      logger.debug { "-- epson Proj, requested to power off" }
+      do_send(:PWR, name: :power_state)
+    end
+  end
+
+  def power?(**options, &block)
+    options[:emit] = block unless block.nil?
+    options[:name] = :power_state
+    do_send(:PWR, **options)
+  end
+
+  #
+  # Input selection
+  #
+  INPUTS = {
+    Inputs::HDMI    => 0x30,
+    Inputs::HDBaseT => 0x80,
+  }
+  INPUT_LOOKUP = INPUTS.invert
+
+  def switch_to(input : Inputs)
+    do_send(:SOURCE, INPUT_LOOKUP[input], name: :inpt_source)
+    do_send(:SOURCE, name: :inpt_query)
+
+    logger.debug { "-- epson LCD, requested to switch to: #{input}" }
+    self[:input] = input # for a responsive UI
+    self[:MUTE] = false
+  end
+
+  #
+  # Volume commands are sent using the inpt command
+  #
+  def volume(vol : Int32, **options)
+    vol = 0 if vol < 0
+    vol = 255 if vol > 255
+
+    # Seems to only return ":" for this command
+    self[:volume] = vol
+    self[:unmute_volume] = vol if vol > 0 # Store the "pre mute" volume, so it can be restored on unmute
+    do_send(:VOL, vol, **options)
+  end
+
+  # Mutes audio + video
+  def mute(
+    state : Bool = true,
+    index : Int32 | String = 0,
+    layer : MuteLayer = MuteLayer::AudioVideo
+  )
+    logger.debug { "-- epson Proj, requested mute state: #{state}" }
+
+    do_send(:MUTE, state, name: :video_mute) # Audio + Video
```

Comment:
You need to honour the `layer` param here.

Reply:
Sorry what do you mean by 'honour'?

Reply:
At the moment if `mute(true, layer: MuteLayer::Audio)` is called, this method will activate the device video mute. You just need to check that param, then execute the appropriate action(s).

Reply:
@dukeraphaelng Something like this https://github.com/PlaceOS/drivers/blob/master/drivers/samsung/displays/mdc_protocol.cr#L124-L132

or for the epson projector example
https://github.com/PlaceOS/drivers/pull/62/commits/60e64c38c5e7d7e967e94d098db0356ba3c55dd7#diff-a9b9e76ba00d89e0a1d6dcd683831bc983f1e69e2ac6736f710d2f4909bb3ca3R110-R116